libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -5344,6 +5347,67 @@ void GCS_MAVLINK::send_received_message_deprecation_warning(const char * message
     send_text(MAV_SEVERITY_WARNING, "Received message (%s) is deprecated", message);
 }
 
+void GCS_MAVLINK::send_altitude() const
+{
+    Location loc;
+    AP_AHRS &_ahrs = AP::ahrs();
+
+    //we wait for a position from AHRS
+    if (!_ahrs.get_position(loc)) {
+        return;
+    }
+    const float altitude_amsl = global_position_int_alt() * 0.001; //meters above ground/sea level
+
+    AP_Baro &_baro = AP::baro();
+    // MAVLink specs for the altitude_monotonic value:
+    // The only guarantee on this field is that it will never be reset and is consistent within a flight.
+    // The recommended value for this field is the uncorrected barometric altitude at boot time.
+    // This altitude will also drift and vary between flights"
+    // Note: ArduPilot allows for ground calibration while disarmed so we can't guarantee (yet) that this field will not be reset!
+    const float altitude_monotonic = _baro.get_altitude();
+
+    Vector3f local_position;
+    // preset to baro altitude and if available use position NED
+    float altitude_local = altitude_monotonic;
+    if (_ahrs.get_relative_position_NED_origin(local_position)) {
+        altitude_local = -local_position.z; //meters from origin
+    }
+
+    // preset to local position and if available use home relative position
+    float altitude_relative = altitude_local;
+    if (_ahrs.home_is_set()) {
+        altitude_relative = global_position_int_relative_alt() * 0.001; //meters above home
+    }
+
+    float altitude_terrain = -1000; //mavlink specs value for unknown
+#if AP_TERRAIN_AVAILABLE
+    AP_Terrain &terrain = AP::terrain();
+    if (terrain.enabled()) {
+        UNUSED_RESULT(terrain.height_above_terrain(altitude_terrain, true));
+    }
+#endif
+
+    float bottom_clearance = -1;    //mavlink specs value for unavailable
+    const RangeFinder *rangefinder = RangeFinder::get_singleton();
+    if (rangefinder != nullptr) {
+        const AP_RangeFinder_Backend *s = rangefinder->find_instance(ROTATION_PITCH_270);
+        if (s != nullptr) {
+            bottom_clearance = s->distance_cm()*0.01;
```

Comment:
If this is out of range somewhere shouldn't we fall back to terrain if we have it, and failing that fall back to relative? I don't see any masking/flags to let a GCS otherwise differentiate on these values.

Reply:
I think so.  I'm actually confused as how this would be different to terrain - except that perhaps terrain wouldn't fall back to relative?

Reply:
Technically terrain should be ground level, while clearance would include things like top of tree, building, whatever rose off the ground below you.